proposals/0148-stake-program-move-instructions.md

```diff
@@ -0,0 +1,201 @@
+
+* `amount` is 0
+* Source or destination are not writable
+* Source or destination are not owned by the stake program
+* Source and destination have the same address
+* Source and destination do not have identical `Authorized`
+* If `Lockup` is in force, source and destination do not have identical `Lockup`
+* The stake account authority is not the `Staker` on both accounts
+* The stake account authority is not a signer
+* Source data length is not equal to the current version of `StakeState`
+* Destination data length is not equal to the current version of `StakeState`
+* Source is not active
+* Destination is neither active nor inactive
+* If destination is active, source and destination are not delegated to the same
+vote account
+* Moving `amount` stake would leave source with a nonzero amount of stake less
+than the minimum delegation
+* Moving `amount` stake would fail to bring destination up to minimum delegation
+
+If all of these conditions hold, then:
+
+* Delegation and lamports on source are debited `amount`
+* Delegation and lamports on destination are credited `amount`
+* If `amount` constitutes the full delegation on the source, source is reset to
+an `Initialized` state
+* If destination is inactive, destination becomes active with the same `Stake`
+as source, aside from delegation amount
+* `credits_observed` must be updated on the destination according to the same
+rules as `Merge`
+
+### `MoveLamports`
+
+`MoveLamports` requires 5 accounts:
+
+0. Source stake account: Writable, owned by the stake program
+1. Destination stake account: Writable, owned by the stake program
+2. Clock: Read-only, with the fixed address
+`SysvarC1ock11111111111111111111111111111111`
+3. Stake history: Read-only, with the fixed address
+`SysvarStakeHistory1111111111111111111111111`
+4. Stake account authority: Read-only signer
+
+`MoveLamports` instruction data is 12 bytes, containing:
+
+* `0x11 0x00 0x00 0x00`, a fixed-value four-byte little-endian unsigned integer
+acting as the instruction discriminator
+* `amount`, an unaligned eight-byte little-endian unsigned integer indicating
+the quantity of lamports to move
+
+`MoveLamports` aborts the transaction when:
+
+* `amount` is 0
+* Source or destination are not writable
+* Source or destination are not owned by the stake program
+* Source and destination have the same address
+* Source and destination do not have identical `Authorized`
+* If `Lockup` is in force, source and destination do not have identical `Lockup`
+* The stake account authority is not the `Staker` on both accounts
+* The stake account authority is not a signer
+* Source is neither active nor inactive
+* Destination is not a valid merge destination (active, inactive, or activating
+with zero effective stake)
+* `amount` exceeds source `lamports - effective_stake - rent_exempt_reserve`
+
+If all of these conditions hold, then:
+
+* Lamports on source are debited `amount`
+* Lamports on destination are credited `amount`
```

Comment:
Does this move the two accounts in activating/deactivating?

Reply:
deactivating isnt allowed for either account, and activating is only allowed for the destination for `MoveLamports`. tbh i describe the allowed states from first principles for the sake of letting the simd stand alone, but a lot of these are just reusing logic from `Merge`, eg here destination is identical to a valid merge destination, and source is a valid merge source except it cant be activating because this would require calculating whether the minimum will be upheld next epoch (which is pointless because theres no usecase)

Reply:
Thank you!